### PR TITLE
Implement last 30 days statistic

### PR DESCRIPTION
**Pull request by VietAnh1010**
Working with date is such is pain.

My current implementation still has some issues. The biggest one right now is how to effectively count the total number of comments.

I will create a draft PR here to keep track of my work.

### CODE REVIEW

**Comment by VietAnh1010** [comment]
Just leave all the conflict to me, don't worry

**Comment by VietAnh1010** [comment]
Hmm, `codecov` fails because diffhit is lower than 96%. Which means I need to add more tests before merging this branch.

I must refractor my code. My code looks like spaghetti right now.

**Comment by joulev** [comment]
image

Very very off-topic, but the word is 'refactor'

**Comment by VietAnh1010** [comment]
@joulev Ready for review.

A few things to note:
- As we now have an dedicated endpoint just for site statistic, `ClientSite` will not contain `SiteStatistic` anymore. But I haven't change this, as changing it could potentially broke front-end at the moment.
- Coverage increased, very nice.

**Comment by joulev** [comment]
Thanks, will have a look tomorrow.

File: server/utils/crud/statisticUtils.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import { firestoreAdmin } from "~/server/firebase/firebaseAdmin";
+import { SITES_COLLECTION } from "~/server/firebase/firestoreCollections";
+
+import { Site, SiteStatistics } from "~/types/server";
+
+import { getDocumentInTransactionWithUid } from "../firestoreUtils";
+import { listSiteComments } from "./siteUtils";
+
+const MILLIS_PER_DAY = 1000 * 60 * 60 * 24;
+
```
Comment (joulev) [inline server/utils/crud/statisticUtils.ts:12]:
```suggestion

function convertTimeToUTCMidnight(timestamp: number) {
    const date = new Date(timestamp);
    date.setUTCHours(0, 0, 0, 0);
    return date.getTime();
}

```

File: server/utils/crud/statisticUtils.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+export async function getSiteStatistic(uid: string, siteId: string) {
+    const siteRef = SITES_COLLECTION.doc(siteId);
+    return await firestoreAdmin.runTransaction(async t => {
+        // Security
+        if (process.env.NODE_ENV !== "development") {
+            await getDocumentInTransactionWithUid<Site>(t, siteRef, uid);
+        }
+        const siteComments = await listSiteComments(siteId);
+        const currentTimestamp = Timestamp.now().toMillis();
```
Comment (joulev) [inline server/utils/crud/statisticUtils.ts:21]:
```suggestion
        const currentTimestamp = convertTimeToUTCMidnight(Timestamp.now().toMillis());
```

File: server/utils/crud/statisticUtils.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        const newComment: number[] = Array.from({ length: 30 }, _ => 0);
+
+        let totalCount = 0;
+        let newCount = 0;
+        let daysAgo = 29;
+
+        for (let i = siteComments.length - 1; i >= 0; i--) {
+            const { date } = siteComments[i];
+            // convert to integer
+            const commentDaysAgo = Math.round((currentTimestamp - date) / MILLIS_PER_DAY);
```
Comment (joulev) [inline server/utils/crud/statisticUtils.ts:32]:
```suggestion
            const commentDaysAgo = Math.round(
                (currentTimestamp - convertTimeToUTCMidnight(date)) / MILLIS_PER_DAY
            );
```

**Review by joulev — COMMENTED** [review]
I think maybe taking the UTC midnight as start of dates makes more sense.

File: server/utils/crud/statisticUtils.ts (hunk 1 above)
Comment (VietAnh1010) [reply server/utils/crud/statisticUtils.ts:12]:
```ts
const toUTCMidnight = (timestamp: number) => timestamp - (timestamp % MILLIS_PER_DAY);
```

This returns the same result, and is faster.

**Pull request merged**